nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -6,35 +6,375 @@
+            // - The device can't be externally reset _during_ this operation because
+            //    the same code responsible for processing the reset request will be
+            //    busy doing the copy.
+            // - We only ever update one RoT stage0 at a time in a rack, so if we brick
+            //    one, only one sled would be affected (still bad).
+            //
+            // So we're ready to roll!
+            debug!(log, "attempting to set RoT bootloader active slot");
+            mgs_clients
+                .try_all_serially(log, move |mgs_client| async move {
+                    let persist = true;
+                    mgs_client
+                        .sp_component_active_slot_set(
+                            update.sp_type,
+                            update.slot_id,
+                            &SpComponent::STAGE0.to_string(),
+                            persist,
+                            &SpComponentFirmwareSlot { slot: 1 },
+                        )
+                        .await?;
+                    Ok(())
+                })
+                .await?;
+
+            debug!(log, "attempting to reset device to set to new RoT bootloader version");
+            mgs_clients
+                .try_all_serially(log, move |mgs_client| async move {
+                    mgs_client
+                        .sp_component_reset(
+                            update.sp_type,
+                            update.slot_id,
+                            &SpComponent::ROT.to_string(),
+                        )
+                        .await?;
+                    Ok(())
+                })
+                .await?;
+
+            Ok(())
+        }
+        .boxed()
+    }
+}
+
+/// Poll the RoT asking for its boot information. This is used to check
+/// state after RoT bootloader updates
+async fn wait_for_stage0_next_image_check(
+    log: &Logger,
+    mgs_clients: &mut MgsClients,
+    sp_type: SpType,
+    sp_slot: u32,
+    timeout: Duration,
+) -> Result<Option<RotImageError>, PostUpdateError> {
+    let before = Instant::now();
+    loop {
+        match mgs_clients
+            .try_all_serially(log, |mgs_client| async move {
+                mgs_client
+                    .sp_rot_boot_info(
+                        sp_type,
+                        sp_slot,
+                        SpComponent::ROT.const_as_str(),
+                        &GetRotBootInfoParams {
+                            version: RotBootInfo::HIGHEST_KNOWN_VERSION,
+                        },
+                    )
+                    .await
+            })
+            .await
+        {
+            Ok(state) => match state.into_inner() {
+                // The minimum we will ever return is 3.
+                // Additionally, V2 does not report image errors, so we cannot
+                // know with certainty if a signature check came back with errors
+                RotState::V2 { .. } => unreachable!(),
```

Comment:
I think we should return a permanent error here instead of panicking. As far as we know we'll only ever see V3 or later, but this is entirely under the control of an external entity; we should not panic if we get unexpected messages from it.

Reply:
Done!